core/src/main/java/org/elasticsearch/indices/IndicesService.java

```diff
@@ -785,39 +785,50 @@ public IndexMetaData verifyIndexIsDeleted(final Index index, final ClusterState
-     *     <li>if the shards data locations do not exists</li>
-     * </ul>
+     * result type returned by {@link #canDeleteShardContent signaling different reasons why a shard can / cannot be deleted}
+     */
+    public enum ShardDeletionCheckResult {
+        CAN_DELETE, // shard data exists and can be deleted
+        STILL_ALLOCATED, // the shard is still allocated / active on this node
+        NO_FOLDER_FOUND, // the shards data locations do not exist
+        SHARED_FILE_SYSTEM, // the shard is located on shared and should not be deleted
+        NO_NODE_LOCK // node does not have filesystem lock to perform lookup
+    }
+
+    /**
+     * Returns <code>ShardDeletionCheckResult</code> signaling whether the shards content for the given shard can be deleted.
      *
      * @param shardId the shard to delete.
      * @param indexSettings the shards's relevant {@link IndexSettings}. This is required to access the indexes settings etc.
      */
-    public boolean canDeleteShardContent(ShardId shardId, IndexSettings indexSettings) {
+    public ShardDeletionCheckResult canDeleteShardContent(ShardId shardId, IndexSettings indexSettings) {
         assert shardId.getIndex().equals(indexSettings.getIndex());
         final IndexService indexService = indexService(shardId.getIndex());
         if (indexSettings.isOnSharedFilesystem() == false) {
            if (nodeEnv.hasNodeFile()) {
                 final boolean isAllocated = indexService != null && indexService.hasShard(shardId.id());
                 if (isAllocated) {
-                    return false; // we are allocated - can't delete the shard
+                    return ShardDeletionCheckResult.STILL_ALLOCATED; // we are allocated - can't delete the shard
                 } else if (indexSettings.hasCustomDataPath()) {
                     // lets see if it's on a custom path (return false if the shared doesn't exist)
                     // we don't need to delete anything that is not there
-                    return Files.exists(nodeEnv.resolveCustomLocation(indexSettings, shardId));
+                    return Files.exists(nodeEnv.resolveCustomLocation(indexSettings, shardId)) ?
+                        ShardDeletionCheckResult.CAN_DELETE :
+                        ShardDeletionCheckResult.NO_FOLDER_FOUND;
                 } else {
                     // lets see if it's path is available (return false if the shared doesn't exist)
                     // we don't need to delete anything that is not there
-                    return FileSystemUtils.exists(nodeEnv.availableShardPaths(shardId));
+                    return FileSystemUtils.exists(nodeEnv.availableShardPaths(shardId)) ?
+                        ShardDeletionCheckResult.CAN_DELETE :
+                        ShardDeletionCheckResult.NO_FOLDER_FOUND;
                 }
-            }
+            } else {
+               return ShardDeletionCheckResult.NO_NODE_LOCK;
```

Comment:
ending up here means the node doesn't have local storage. Can we rename this enum? (see DiscoveryNode#nodeRequiresLocalStorage and NodeEnvironment#NodeEnvironment)